app/user/user_main.c

```diff
@@ -47,6 +47,49 @@ void TEXT_SECTION_ATTR user_start_trampoline (void)
   rtctime_early_startup ();
 #endif
 
+  /* Re-implementation of default init data deployment. The SDK does not
+   * appear to be laying down its own version of init data anymore, so
+   * we have to do it again. To see whether we need to, we read out
+   * the flash size and do a test for esp_init_data based on that size.
+   * If it's missing, we need to initialize it *right now* before the SDK
+   * starts up and gets stuck at "rf_cal[0] !=0x05,is 0xFF".
+   * If the size byte is wrong, then we'll end up fixing up the init data
+   * again on the next boot, after we've corrected the size byte.
+   * Only remaining issue is lack of spare code bytes in iram, so this
+   * is deliberately quite terse and not as readable as one might like.
+   */
+  SPIFlashInfo sfi;
+  SPIRead (0, (uint32_t *)(&sfi), sizeof (sfi)); // Cache read not enabled yet, safe to use
+  if (sfi.size < 2) // Compensate for out-of-order 4mbit vs 2mbit values
+    sfi.size ^= 1;
+  uint32_t flash_end_addr = (256 * 1024) << sfi.size;
+  uint32_t init_data_hdr = 0xffffffff;
+  uint32_t init_data_addr = flash_end_addr - 4 * SPI_FLASH_SEC_SIZE;
+  SPIRead (init_data_addr, &init_data_hdr, sizeof (init_data_hdr));
+  if (init_data_hdr == 0xffffffff)
+  {
+    // Note: ideally we should generate the actual init data from the
+    // SDK's bin/esp_init_data_default.bin during the build, and #include it
+    // straight into this array. E.g.:
+    //   xxd -i esp_init_data_default | tail -n13 | head -n11 > eidd.h
+    static uint8_t flash_init_data[128] __attribute__((aligned(4))) =
```

Comment:
If this is indeed what I think it is then it's the same I complained about in #1500. The init data bytes shouldn't be hardcoded (will soon be outdated) but read from the SDK which is downloaded during the build. @vowstar fixed this in his PR with https://github.com/nodemcu/nodemcu-firmware/pull/1500/commits/7f3e4b8257875b3786b2793cff59bd0ecd44aa93.